### PR TITLE
CreateShipment: Create Page and Test Cases

### PR DESCRIPTION
**Pull request by vahidbazzaz**
This PR adds the ability to create shipments and also includes test cases 4.3 based on the test plan

### CODE REVIEW

**Comment by vahidbazzaz** [comment]
[2.0 CreateShipment: write tests](https://trello.com/c/kicxMHmk)
 [2.0 CreateShipment: create page](https://trello.com/c/XspMphto)

File: back/boxtribute_server/business_logic/box_transfer/agreement/crud.py [hunk 1]
```diff
@@ -179,6 +179,6 @@ def retrieve_transfer_agreement_bases(*, transfer_agreement, kind):
         TransferAgreementDetail, on=getattr(TransferAgreementDetail, f"{kind}_base")
     ).where(
         TransferAgreementDetail.transfer_agreement == transfer_agreement.id
-    ) or Base.select().where(
+    ).distinct() or Base.select().where(
```
Comment (vahidbazzaz) [inline back/boxtribute_server/business_logic/box_transfer/agreement/crud.py:182]: @pylipp suggested this change to resolve the redundant targetBases

**Review by pylipp — COMMENTED** [review]
Functional review
:heavy_check_mark: after having created and accepted an agreement, I could start a new shipment
:heavy_check_mark: user cannot submit data if input incomplete

File: react/src/mocks/transferAgreements.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
+export const transferAgreement1 = {
```
Comment (pylipp) [inline react/src/mocks/transferAgreements.ts:1]: maybe call it `acceptedTransferAgreement`?

File: react/src/queries/fragments.ts [hunk 3]
```diff
@@ -155,3 +155,51 @@ export const BOX_WITH_SIZE_TAG_PRODUCT_FIELDS_FRAGMENT = gql`
     }
   }
 `;
+
+export const TRANSFER_AGREEMENT_FEILDS_FRAGMENT = gql`
```
Comment (pylipp) [inline react/src/queries/fragments.ts:159]: typo: FIELDS

File: react/src/queries/fragments.ts [hunk 4]
```diff
@@ -155,3 +155,51 @@ export const BOX_WITH_SIZE_TAG_PRODUCT_FIELDS_FRAGMENT = gql`
...
+    }
+    targetOrganisation {
+      id
+      name
+    }
+    targetBases {
+      id
+      name
+    }
+    type
```
Comment (pylipp) [inline react/src/queries/fragments.ts:178]: duplicated, same for state and validFrom

File: react/src/views/Transfers/CreateTransferShipment/CreateTransferShipmentView.test.tsx [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+  const submitStartButton = await screen.findByRole("button", { name: /start/i });
+  expect(submitStartButton).toBeInTheDocument();
+
+  await assertOptionsInSelectField(user, /organisation/i, [/boxcare/i], pageTitle);
+  await selectOptionInSelectField(user, /organisation/i, "BoxCare");
+  expect(await screen.findByText("BoxCare")).toBeInTheDocument();
+  await selectOptionInSelectField(user, /base/i, "Samos");
+  expect(await screen.findByText("Samos")).toBeInTheDocument();
+  await user.click(submitStartButton);
+  expect(await screen.findByText(/your changes could not be saved!/i)).toBeInTheDocument();
```
Comment (pylipp) [inline react/src/views/Transfers/CreateTransferShipment/CreateTransferShipmentView.test.tsx:209]: I added a new test case 4.3.3.4 to the test plan. It's similar to 4.3.3.3 but the `errors` list in the mutation response needs to be filled out (to mock an error in the BE during processing the request). Can you implement this?

File: react/src/views/Transfers/CreateTransferShipment/CreateTransferShipmentView.tsx [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    CreateTransferShipmentMutationVariables
+  >(CREATE_SHIPMENT_MUTATION);
+
+  // Prep data for Form
+  const allTransferAgreements = allFormOptions.data?.transferAgreements;
+
+  // eslint-disable-next-line max-len
+  const currentOrganisationLabel = `${allFormOptions?.data?.base?.organisation?.name} - ${allFormOptions?.data?.base?.name}`;
+
+  const partnerOrgsAgreementData = allTransferAgreements
```
Comment (pylipp) [inline react/src/views/Transfers/CreateTransferShipment/CreateTransferShipmentView.tsx:86]: @vahidbazzaz can you explain what happens here? The list of all accepted agreements is filtered for these that have the user's org as source and that have types SendingTo or Bidirectional.
But it should be _all_ accepted agreements that have the user's org as source, independent of the type (shipments are always created by the source side).

File: react/src/views/Transfers/CreateTransferShipment/CreateTransferShipmentView.tsx [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+          orgId: agreement.targetOrganisation.id,
+          orgName: agreement.targetOrganisation.name,
+          orgBases: agreement.targetBases,
+        } as IOrganisationsAgreementsDataData),
+    );
+
+  // Handle Submission
+  const onSubmitCreateShipmentForm = (createTransferShipmentData: ITransferShipmentFormData) => {
+    // Choose the agreement id based on the partner organization's base id
+    const selectedAgreement = partnerOrgsAgreementData?.find((org) =>
```
Comment (pylipp) [inline react/src/views/Transfers/CreateTransferShipment/CreateTransferShipmentView.tsx:107]: I'm confused by the variable `org`, doesn't `partnerOrgsAgreementData` contain agreements?

File: react/src/views/Transfers/CreateTransferShipment/CreateTransferShipmentView.tsx [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+      (agreement) =>
+        agreement.sourceOrganisation.id === globalPreferences.selectedOrganisationId?.toString() &&
+        (agreement.type === TransferAgreementType.SendingTo ||
+          agreement.type === TransferAgreementType.Bidirectional),
+    )
+    .map(
+      (agreement) =>
+        ({
+          agreementId: agreement.id,
+          specialNote: agreement.comment,
```
Comment (pylipp) [inline react/src/views/Transfers/CreateTransferShipment/CreateTransferShipmentView.tsx:97]: This field, and orgId and orgName are not used elsewhere if I'm not mistaken. Can we remove them here?

Comment (pylipp) [reply react/src/views/Transfers/CreateTransferShipment/CreateTransferShipmentView.tsx:97]: NVM I see that they're used in the low-level component.

**Comment by pylipp** [comment]
Code review also done with a couple of open questions. I think we have to talk about how the create-shipment page is reached. If it's via the agreement overview then we already know which agreement the shipment should be created for. Otherwise if there are multiple agreements between two orgs, it should maybe be indicated in the org-select dropdown (`OrgName (agreement #1)`, `OrgName (agreement #3)`). @aerinsol not sure if this is already being considered in this design.

Also @HaGuesto would appreciate if you do a code review.

File: react/src/views/Transfers/CreateTransferShipment/CreateTransferShipmentView.tsx (hunk 6 above)
Comment (pylipp) [reply react/src/views/Transfers/CreateTransferShipment/CreateTransferShipmentView.tsx:86]: @vahidbazzaz added a (hopefully) clarifying note above :upside_down_face:

File: react/src/views/Transfers/CreateTransferShipment/CreateTransferShipmentView.tsx (hunk 7 above)
Comment (vahidbazzaz) [reply react/src/views/Transfers/CreateTransferShipment/CreateTransferShipmentView.tsx:107]: @pylipp In the shipment UI, we just have the partner org's base information, and we need to match the selected partner org's base id to the corresponding agreement id.

Comment (vahidbazzaz) [reply react/src/views/Transfers/CreateTransferShipment/CreateTransferShipmentView.tsx:107]: @pylipp I added optional data props to the SelectField so we can add agreement data as data and remove extra logic to retrieve agreements

File: react/src/components/Form/SelectField.tsx [hunk 9]
```diff
@@ -7,6 +7,7 @@ export interface IDropdownOption extends OptionBase {
   value: string;
   label: string;
   color?: string | undefined | null;
+  data?: object | undefined | null;
```
Comment (HaGuesto) [inline react/src/components/Form/SelectField.tsx:10]: the QuestionMark should be enough. not sure why you need to define the case `| undefined | null`
btw it is bad practise to use null. I try mostly to only use undefined. This can be sometimes tough though.

https://medium.com/@o_o_o/null-vs-undefined-can-i-use-only-one-a3b7db5468f2

File: react/src/queries/fragments.ts [hunk 10]
```diff
@@ -155,3 +155,49 @@ export const BOX_WITH_SIZE_TAG_PRODUCT_FIELDS_FRAGMENT = gql`
...
+    state
+    shipments {
+      id
+      state
+    }
+    type
```
Comment (HaGuesto) [inline react/src/queries/fragments.ts:184]: type is still duplicate

File: react/src/views/Transfers/CreateTransferShipment/CreateTransferShipmentView.test.tsx [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+import { acceptedTransferAgreement } from "mocks/transferAgreements";
+import userEvent from "@testing-library/user-event";
+import { assertOptionsInSelectField, selectOptionInSelectField } from "tests/helpers";
+import { GraphQLError } from "graphql";
+import CreateTransferShipmentView, {
+  ALL_ORGS_AND_BASES_WITH_TRANSFER_AGREEMENTS_QUERY,
+  CREATE_SHIPMENT_MUTATION,
+} from "./CreateTransferShipmentView";
+
+jest.setTimeout(30000);
```
Comment (HaGuesto) [inline react/src/views/Transfers/CreateTransferShipment/CreateTransferShipmentView.test.tsx:13]: This is scary. Why is this needed?

File: react/src/views/Transfers/CreateTransferShipment/CreateTransferShipmentView.test.tsx [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+  expect(await screen.findByText("Samos")).toBeInTheDocument();
+
+  await user.click(submitButton);
+  expect(await screen.findByText(/successfully created a new shipment/i)).toBeInTheDocument();
+  // Test case 4.3.3.2 - Redirect to Transfers Shipments Page
+  expect(
+    await screen.findByRole("heading", { name: "/bases/1/transfers/shipments" }),
+  ).toBeInTheDocument();
+
+  cleanup();
```
Comment (HaGuesto) [inline react/src/views/Transfers/CreateTransferShipment/CreateTransferShipmentView.test.tsx:181]: Why do you not split this up into an extra test?

File: react/src/views/Transfers/CreateTransferShipment/CreateTransferShipmentView.tsx [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+              type: "success",
+              message: "Successfully created a new shipment",
+            });
+
+            navigate(`/bases/${baseId}/transfers/shipments`);
+          }
+        })
+        .catch((err) => {
+          triggerError({
+            message: "Your changes could not be saved!",
```
Comment (HaGuesto) [inline react/src/views/Transfers/CreateTransferShipment/CreateTransferShipmentView.tsx:138]: Prob the message should be different, like `Error while trying to create a new shipment`

File: react/src/views/Transfers/CreateTransferShipment/components/CreateTransferShipment.tsx [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+import { Props } from "chakra-react-select";
+import { useNavigate, useParams } from "react-router-dom";
+
+export interface IBaseData {
+  id: string;
+  name: string;
+}
+
+// eslint-disable-next-line react/function-component-definition
+export const ReceivingIcon: FunctionComponent<Props> = () => (
```
Comment (HaGuesto) [inline react/src/views/Transfers/CreateTransferShipment/components/CreateTransferShipment.tsx:30]: can u define them somewhere globally? Then I can also use them

File: react/src/views/Transfers/CreateTransferShipment/components/CreateTransferShipment.tsx [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+    <path
+      // eslint-disable-next-line max-len
+      d="M12.172 11.778L6.808 6.414L8.222 5L16 12.778L8.222 20.556L6.808 19.142L12.172 13.778H0V11.778H12.172Z"
+      fill="#2D3748"
+    />
+  </Icon>
+);
+
+// eslint-disable-next-line react/function-component-definition
+export const SendingIcon: FunctionComponent<Props> = () => (
```
Comment (HaGuesto) [inline react/src/views/Transfers/CreateTransferShipment/components/CreateTransferShipment.tsx:42]: can u define them somewhere globally? Then I can also use them

File: react/src/views/Transfers/CreateTransferShipment/components/CreateTransferShipment.tsx [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+            <SelectField
+              fieldId="partnerOrganisationSelectedBase"
+              fieldLabel="Base"
+              placeholder="Please select a base"
+              errors={errors}
+              control={control}
+              options={basesOptionsForPartnerOrg}
+            />
+          </ListItem>
+          {/* <ListItem>
```
Comment (HaGuesto) [inline react/src/views/Transfers/CreateTransferShipment/components/CreateTransferShipment.tsx:215]: remove

File: react/src/views/Transfers/CreateTransferShipment/components/CreateTransferShipment.tsx [hunk 17]
```diff
@@ -0,0 +1,254 @@
...
+
+            <Button
+              mt={10}
+              isLoading={isSubmitting}
+              type="submit"
+              borderRadius="0"
+              w="full"
+              variant="solid"
+              backgroundColor="blue.500"
+              color="white"
```
Comment (HaGuesto) [inline react/src/views/Transfers/CreateTransferShipment/components/CreateTransferShipment.tsx:243]: why white color?

**Review by HaGuesto — COMMENTED** [review]
I did a quick code review. Looks quite good. Just a few little things to improve.

File: react/src/views/Transfers/CreateTransferShipment/components/CreateTransferShipment.tsx [hunk 18]
```diff
@@ -0,0 +1,254 @@
...
+  partnerOrganisation: singleSelectOptionSchema
+    .refine(Boolean, { message: "Please select a partner organisation" })
+    .transform((selectedOption) => selectedOption || { label: "", value: "" }),
+  partnerOrganisationSelectedBase: singleSelectOptionSchema
+    .refine(Boolean, { message: "Please select a partner base" })
+    .transform((selectedOption) => selectedOption || { label: "", value: "" }),
+});
+
+export type ITransferShipmentFormData = z.infer<typeof TransferShipmentFormDataSchema>;
+
```
Comment (HaGuesto) [inline react/src/views/Transfers/CreateTransferShipment/components/CreateTransferShipment.tsx:82]: I would add here the definition of default value for partnerOrganisation and partnerOrganisationSelectedBase if we redirect from AgreementsOverview

Comment (HaGuesto) [reply react/src/views/Transfers/CreateTransferShipment/components/CreateTransferShipment.tsx:82]: https://reactrouter.com/en/main/components/link#state

Comment (HaGuesto) [reply react/src/views/Transfers/CreateTransferShipment/components/CreateTransferShipment.tsx:82]: In the navigate function I would then add a state parameter.

File: react/src/views/Transfers/CreateTransferShipment/CreateTransferShipmentView.tsx (hunk 13 above)
Comment (vahidbazzaz) [reply react/src/views/Transfers/CreateTransferShipment/CreateTransferShipmentView.tsx:138]: @HaGuesto Sure, I'll update the test plan error message text and change the message.

File: react/src/views/Transfers/CreateTransferShipment/CreateTransferShipmentView.test.tsx (hunk 11 above)
Comment (vahidbazzaz) [reply react/src/views/Transfers/CreateTransferShipment/CreateTransferShipmentView.test.tsx:13]: @HaGuesto It resolves the random timeout issue when CircleCI runs the tests, but we can remove it locally. If you have any other ideas, please let me know.

File: react/src/views/Transfers/CreateTransferShipment/components/CreateTransferShipment.tsx (hunk 17 above)
Comment (vahidbazzaz) [reply react/src/views/Transfers/CreateTransferShipment/components/CreateTransferShipment.tsx:243]: @HaGuesto The UI should display the submit button text color in white, but I think we can have a general submit button within the theme so we don't need to apply style to it

**Comment by HaGuesto** [comment]
@pylipp 

> I think we have to talk about how the create-shipment page is reached. If it's via the agreement overview then we already know which agreement the shipment should be created for. 
We do not have a direct redirect from AgreementsOverview to CreateShipment according to the design atm.

> Otherwise if there are multiple agreements between two orgs, it should maybe be indicated in the org-select dropdown (OrgName (agreement #1), OrgName (agreement #3)). @aerinsol not sure if this is already being considered in this design.

There was a discussion about this on monday (6th of Feb)

**Comment by aerinsol** [comment]
We agreed for time being there should be no agreement duplicates for the same base @pylipp

**Review by pylipp — APPROVED** [review]
